### PR TITLE
✨ delist button (unlist)

### LINKED ISSUE #4819: Delist button missing
delist button is missing when you have listed nfts and you are the owner. 
image

for burn button i think there is already created issue.

### PR DESCRIPTION
**Pull request by petersopko**
## PR Type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

- [x] Closes #4819
- [ ] Requires deployment <>

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [x] I've tested it at </bsx/collection>
- [x] I've tested PR on mobile
- [x] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=FNyt7T1xdbhN7dagf7yGYCRJuE3R45VmTCE3tjzy1rKxa7y)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

rmrk:
```
{
  "data": {
    "nftEntityById": {
      "blockNumber": "13845250",
      "id": "13845250-A4598F5615F201D739-UDF-012-0000000000000012",
      "name": "012",
      "events": [
        {
          "id": "13845250-A4598F5615F201D739-UDF-012-0000000000000012-MINTNFTSR3HEF6npq_u-_zPHdSr9",
          "timestamp": "2022-08-03T13:17:00.017000Z",
          "interaction": "MINTNFT"
        },
        {
          "id": "13845250-A4598F5615F201D739-UDF-012-0000000000000012-LISTFawQAgOlj8N_T8u5RnVGk",
          "timestamp": "2022-08-03T13:17:42.019000Z",
          "interaction": "LIST"
        },
        {
          "id": "13845250-A4598F5615F201D739-UDF-012-0000000000000012-LISTbfmu9snljHJoWu6D6Vv_r",
          "timestamp": "2022-08-08T11:56:30.031000Z",
          "interaction": "LIST"
        },
        {
          "id": "13845250-A4598F5615F201D739-UDF-012-0000000000000012-UNLISTkcRXdMrRk_VO7H_pp-m7J",
          "timestamp": "2023-01-25T07:13:18.000000Z",
          "interaction": "UNLIST"
        }
      ]
    }
  }
}
```

snek:
```
{
  "data": {
    "nftEntityById": {
      "blockNumber": "776899",
      "id": "3424749042-2",
      "name": "seduction",
      "events": [
        {
          "interaction": "MINTNFT",
          "id": "3424749042-2-MINTNFT--B6Q4iucHFO-63IY9IRZY",
          "timestamp": "2022-09-07T04:40:48.240000Z"
        },
        {
          "interaction": "LIST",
          "id": "3424749042-2-LIST-tcd16Xnr7M5OW0kDWFWca",
          "timestamp": "2022-09-07T04:40:48.240000Z"
        },
        {
          "interaction": "SEND",
          "id": "3424749042-2-SEND-M3UWnXlwBRiZZ9tL58iT9",
          "timestamp": "2022-09-23T16:22:36.121000Z"
        },
        {
          "interaction": "BUY",
          "id": "3424749042-2-BUY-bnRA92odx04esW8b-UR7y",
          "timestamp": "2022-09-23T16:22:36.121000Z"
        },
        {
          "interaction": "UNLIST",
          "id": "3424749042-2-UNLIST-JU1_nySFVK_Q4rCjS5f4M",
          "timestamp": "2023-01-25T06:14:42.121000Z"
        },
        {
          "interaction": "LIST",
          "id": "3424749042-2-LIST-MKA-dS5TUZEXLaBZxu7fd",
          "timestamp": "2023-01-25T07:23:36.147000Z"
        },
        {
          "interaction": "UNLIST",
          "id": "3424749042-2-UNLIST-xGwEgb_XejhkNaGMOVhPh",
          "timestamp": "2023-01-25T07:24:42.150000Z"
        }
      ]
    }
  }
}
```

### CODE REVIEW

**Review by roiLeo — APPROVED** [review]
✅ code lgtm

IDK if it's needed to add "UNLIST" interaction maybe instead do something like:
```
const unlist = () => {
  transaction({
    interaction: Interaction.LIST,
    urlPrefix: urlPrefix.value,
    price: 0,
    nftId: $route.params.id,
    successMessage: $i18n.t('transaction.unlist.success') as string,
    errorMessage: $i18n.t('transaction.unlist.error') as string,
  })
}
```

**Comment by petersopko** [comment]
> IDK if it's needed to add "UNLIST" interaction

I'm not sure of the convention here, but I was trying to use the enum in minimark:

```
export enum Interaction {
  MINT = 'MINT',
  MINTNFT = 'MINTNFT',
  LIST = 'LIST',
  UNLIST = 'UNLIST',
  BUY = 'BUY',
  SEND = 'SEND',
  CONSUME = 'CONSUME',
  CHANGEISSUER = 'CHANGEISSUER',
  EMOTE = 'EMOTE',
}
```

I can remove it if this isn't what we want.

**Comment by roiLeo** [comment]
> I can remove it if this isn't what we want.

I'm fine with that, "UNLIST" event doesn't exist on chain, it's just "LIST" with price 0

**Comment by petersopko** [comment]
> I'm fine with that, "UNLIST" event doesn't exist on chain, it's just "LIST" with price 0

That is true, but historically speaking, we've been using that as a datapoint (for example in item history). You can filter for UNLIST events in indexer.

Maybe we don't want to get rid of that? Idk, wdyt @vikiival ?

I'll do whatever is needed, I'm just not sure about the effects of removing UNLIST now.

**Comment by vikiival** [comment]
> Maybe we don't want to get rid of that? Idk, wdyt @vikiival ?

That sound ambigous in the context, can you please clarify?

But yeah unlist is list for 0
And I would like to keep it that way because for example Astar will have possibility to claim free NFTs

**Comment by petersopko** [comment]
nevermind, I fundamentally misunderstood the way these transactions interact with indexer.

I simplified the code as @roiLeo suggested, still works as intended, i.e. it's shown as unlist 👍🏻 
image


```
{
  "data": {
    "events": [
      {
        "interaction": "UNLIST",
        "timestamp": "2023-01-25T11:27:30.000000Z",
        "id": "13845059-A4598F5615F201D739-UDF-004-0000000000000004-UNLISTK0FcBUmf8OVHvNGByZYQu"
      },
      {
        "interaction": "LIST",
        "timestamp": "2022-08-08T11:55:00.015000Z",
        "id": "13845059-A4598F5615F201D739-UDF-004-0000000000000004-LIST3xAm1QE558QnUtU3Z3Ioc"
      },
      {
        "interaction": "LIST",
        "timestamp": "2022-08-03T12:58:30.015000Z",
        "id": "13845059-A4598F5615F201D739-UDF-004-0000000000000004-LIST8okhjw_zMdwXMQPanY4La"
      },
      {
        "interaction": "MINTNFT",
        "timestamp": "2022-08-03T12:57:54.008000Z",
        "id": "13845059-A4598F5615F201D739-UDF-004-0000000000000004-MINTNFTFPggjunjyJYw7sdKlWEei"
      }
    ]
  }
}
```

**Comment by petersopko** [comment]
Ok, there's actually an issue here, I guess the previous comment was caused by hot reload bug, can't successfully reproduce it :D 

Here's the check we're doing at list:
```
function checkTsxList(item: ActionList) {
  const meta = Number(item.price)

  if (!meta) {
    dangerMessage('Price is not valid')
    return false
  }

  return true
}
```

Do we want to somehow go around it in this case? I guess the `Interaction.LIST` shouldn't be allowed with price `0`?

**Comment by roiLeo** [comment]
> Do we want to somehow go around it in this case? I guess the `Interaction.LIST` shouldn't be allowed with price `0`?

Maybe we want to change `checkTsxList` to check negative price instead?

**Comment by petersopko** [comment]
> edit: I don't really see the point of this check

Yeah, I removed the check and got to this:
```
"interaction": "LIST",
"id": "13844963-A4598F5615F201D739-UDF-001-0000000000000001-LISTtiJtAdEw3-kZ2fmBsZJ1a",
"timestamp": "2023-01-25T16:00:54.000000Z",
"meta": "-14000000000000"
```

image

I don't think we want to allow that, so I've set the check to catch negative numbers only, should be all ok now.

**Comment by roiLeo** [comment]
> I don't think we want to allow that, so I've set the check to catch negative numbers only, should be all ok now.

wut? is it on rmrk or bsx? I thought there was at least something to check negative values on nft pallet level

**Comment by roiLeo** [comment]
Tested negative valu on bsx and it's not possible
```bash
Error: createType(Call):: Call: failed decoding marketplace.setPrice:: Struct: failed on args: {"collection_id":"u128","item_id":"u128","new_price":"Option"}:: Struct: failed on new_price: Option:: u128: Negative number passed to unsigned type
```

File: composables/transaction/transactionList.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ import type { ActionList } from './types'
 function checkTsxList(item: ActionList) {
```
Comment (vikiival) [inline composables/transaction/transactionList.ts:8]:
```suggestion
function isListTxValid(item: ActionList) {
```

**Review by vikiival — CHANGES_REQUESTED** [review]
smol

File: composables/transaction/transactionList.ts (hunk 1 above)
Comment (petersopko) [reply composables/transaction/transactionList.ts]: done

**Review by vikiival — CHANGES_REQUESTED** [review]
Last piece

File: components/gallery/GalleryItemButton/GalleryItemMoreActionBtn.vue [hunk 2]
```diff
@@ -1,19 +1,26 @@
...
+          @click.native="downloadMedia" />
+        <NeoDropdownItem
+          v-if="currentOwner === accountId"
+          item="Burn"
+          @click.native="burn" />
+        <NeoDropdownItem
+          v-if="accountId === currentOwner && price !== '0'"
+          item="Delist"
+          @click.native="unlist" />
+        <NeoDropdownItem disabled item="Report" />
```
[the remark this answers is not in the document]
Comment (petersopko) [reply components/gallery/GalleryItemButton/GalleryItemMoreActionBtn.vue]: done

**Review by vikiival — APPROVED** [review]
LGTM

**Comment by vikiival** [comment]
pay 35 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $35 @ 35.17 USD/KSM ~ 0.995 $KSM
🧗 [FNyt7T1xdbhN7dagf7yGYCRJuE3R45VmTCE3tjzy1rKxa7y](https://kusama.subscan.io/account/FNyt7T1xdbhN7dagf7yGYCRJuE3R45VmTCE3tjzy1rKxa7y)
🔗 [0x0d4b325c5985a70fabbd217b73fcbf405980efb74fd128eb74e32d9150510eb3](https://kusama.subscan.io/extrinsic/0x0d4b325c5985a70fabbd217b73fcbf405980efb74fd128eb74e32d9150510eb3)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)
